Unify genclient tags and add more fine control on verbs generated

mfojtik opened the pull request:
This will change the syntax of the existing `genclient` tags be like this:

```
// +genclient
// +genclient:noStatus
// +genclient:noVerbs
// +genclient:nonNamespaced
// +genclient:readonly
```

The first one indicates the client will be generated from the struct below and the other tags are basically options to the genclient (which justify why they should be prefixed with `genclient:`)

This also changes the `// +genclientstatus=false` to `// +genclient:noStatus` to follow the pattern and also changes the `// +noMethods=true` to `// +genclient:noVerbs` as we call the REST operations verbs so it will make it consistent with terminology.

In addition to existing options this patch also add two more to allow more fine-grained control on which verbs are going to be generated. This is extra useful for third-party projects (like OpenShift) where some resources does not implement full CRUD, but for example just "create" verb or "create" and "delete"...
To support that, you can use this syntax:

```
// +genclient:onlyVerbs=create,delete
// +genclient:skipVerbs=patch
```

The first one will generate only create and delete functions and second one will generate full CRUD without "patch" actions. This somehow overlaps with the existing "readonly" tag, but I want to keep that tag in place as it reads better in some cases ;-)

mfojtik committed:
unify tag syntax for genclient tags and add onlyVerbs and skipVerbs

mfojtik committed:
update tags in types for new genclient syntax

mfojtik committed:
regenerate clients to pickup updated genclient:noStatus comment

mfojtik commented:
@deads2k @sttts PTAL 

This is based on discussion in https://github.com/openshift/origin/pull/15236

sttts commented:
Looks good.

sttts commented:
/cc @kubernetes/sig-api-machinery-api-reviews

mfojtik committed:
update staging clients

deads2k commented on cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go:
```diff
@@ -148,49 +152,85 @@ func (g *genFakeForType) GenerateType(c *generator.Context, t *types.Type, w io.
 		sw.Do(resource, m)
 		sw.Do(kind, m)
 	}
 
```
Can we actually return here if its `tags.NoVerbs`?

deads2k commented on cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go:
```diff
@@ -148,49 +152,85 @@ func (g *genFakeForType) GenerateType(c *generator.Context, t *types.Type, w io.
 
-	if !noMethods && !readonly {
-		sw.Do(patchTemplate, m)
+	if !tags.NoVerbs && !tags.Readonly {
```
why is this if statement appearing again?  It looks the same as above

mfojtik replied:
yeah, it makes sure the Patch() will be the last method (i can move it upper but that will reorder the methods... not sure if we care about that)

deads2k commented on cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go:
```diff
@@ -148,49 +152,85 @@ func (g *genFakeForType) GenerateType(c *generator.Context, t *types.Type, w io.
-			sw.Do(listUsingOptionsTemplate, m)
-		} else {
-			sw.Do(listTemplate, m)
+	if !tags.NoVerbs {
```
returning early makes this block unconditional and early.

deads2k commented on cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go:
```diff
@@ -148,49 +152,85 @@ func (g *genFakeForType) GenerateType(c *generator.Context, t *types.Type, w io.
-		sw.Do(updateTemplate, m)
-		// Generate the UpdateStatus method if the type has a status
-		if genStatus(t) {
+	if !tags.NoVerbs && !tags.Readonly {
```
After you generate getters, you can check if its readonly and just return, right?

deads2k commented on cmd/libs/go2idl/client-gen/generators/generator_for_type.go:
```diff
@@ -95,74 +98,133 @@ func (g *genClientForType) GenerateType(c *generator.Context, t *types.Type, w i
-	noMethods := extractBoolTagOrDie("noMethods", t.SecondClosestCommentLines) == true
 
-	readonly := extractBoolTagOrDie("readonly", t.SecondClosestCommentLines) == true
+	if len(tags.OnlyVerbs) > 0 {
```
This block looks familiar. Why wouldn't this be done in the parse step?

deads2k commented on cmd/libs/go2idl/client-gen/generators/generator_for_type.go:
```diff
@@ -95,74 +98,133 @@ func (g *genClientForType) GenerateType(c *generator.Context, t *types.Type, w i
 	return sw.Error()
 }
 
+func skipVerb(m string, skippedVerbs []string) bool {
```
this is starting to look like a helper on the tags.  Second time I've seen it.

mfojtik replied:
yeah I will move this to util

deads2k commented on cmd/libs/go2idl/lister-gen/generators/expansion.go:
```diff
@@ -41,10 +43,14 @@ func (g *expansionGenerator) Filter(c *generator.Context, t *types.Type) bool {
 	for _, t := range g.types {
+		tags, err := util.ParseClientGenTags(t.SecondClosestCommentLines)
+		if err != nil {
+			fmt.Printf("Error parsing tags: %v", err)
```
if we can't parse, our return code should be non-zero.  This seems to hide that.

deads2k commented:
I'm definitely in favor.  I have a few minor comments and one concern (spread throughout the pull) about error handling.

mfojtik commented:
@deads2k addressed all comments which also caused all clients to regenerate and reorder the functions :-) somebody should probably double-check that the generated clients are correct as I can have bug in the logic but it is hard to tell :-)

deads2k commented on cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_type.go:
```diff
@@ -152,85 +148,59 @@ func (g *genFakeForType) GenerateType(c *generator.Context, t *types.Type, w io.
 	}
 
-	if !tags.NoVerbs {
+	if tags.NoVerbs {
```
This conditional switched.  Need an if/else?

mfojtik replied:
i see

deads2k commented:
One comment on an if/else block.  otherwise lgtm.

mfojtik commented:
@deads2k updated and regenerated

deads2k commented:
@sttts ready for another pass I think.

mfojtik commented:
fixed unit test
/retest

spxtr commented:
/lint

mfojtik commented:
@sttts rebased on top of your PR, pls PTAL, every change to API from now causing this to conflict...

deads2k commented:
/lgtm
/approve no-issue

deads2k commented:
outstanding approvals are the tag updates to be consistent.  tagging.

sttts commented on staging/src/k8s.io/kube-gen/cmd/client-gen/generators/generator_for_type.go:
```diff
@@ -95,74 +93,108 @@ func (g *genClientForType) GenerateType(c *generator.Context, t *types.Type, w i
 	}
 
-	if !noMethods && !readonly {
+	if tags.Readonly {
```
I hope this does not mean you have to manually remove the writing verbs?

mfojtik replied:
@sttts this should just skip generating them (return below).

sttts commented on staging/src/k8s.io/kube-gen/cmd/client-gen/util/tags.go:
```diff
@@ -0,0 +1,132 @@
+		ret.SkipVerbs = strings.Split(v[0], ",")
+	}
+	if v, exists := values["genclient:onlyVerbs"]; exists {
+		onlyVerbs := strings.Split(v[0], ",")
```
there should be a conflict if skipVerbs and onlyVerbs both are used

sttts commented on staging/src/k8s.io/kube-gen/cmd/client-gen/util/tags.go:
```diff
@@ -0,0 +1,132 @@
+	_, ret.NonNamespaced = values["genclient:nonNamespaced"]
+	_, ret.NoVerbs = values["genclient:noVerbs"]
+	_, ret.NoStatus = values["genclient:noStatus"]
+	_, ret.Readonly = values["genclient:readonly"]
```
readonly should set SkipVerbs

sttts commented on staging/src/k8s.io/kube-gen/cmd/client-gen/util/tags_test.go:
```diff
@@ -0,0 +1,64 @@
+			expectTags: Tags{GenerateClient: true, NoVerbs: true},
+		},
+		"genclient:noStatus": {
+			lines:      []string{`+genclient`, `+genclient:noStatus`},
```
readonly missing in the tests

sttts replied:
skipVerbs+onlyVerbs at the same time are missing.

sttts commented on staging/src/k8s.io/kube-gen/cmd/client-gen/util/tags_test.go:
```diff
@@ -0,0 +1,64 @@
+		expectError bool
+	}{
+		"genclient": {
+			lines:      []string{`+genclient`},
```
what happens with `+genclient=true`. Gives an error?

mfojtik replied:
@sttts no, it is supported (just checking the existance of key) would you expect an error?

sttts replied:
just make sure that `+genclient=false` is not confused with `+genclient`.

sttts commented on staging/src/k8s.io/kube-gen/cmd/client-gen/generators/fake/generator_fake_for_group.go:
```diff
@@ -70,18 +71,20 @@ func (g *genFakeForGroup) GenerateType(c *generator.Context, t *types.Type, w io
 
 	sw.Do(groupClientTemplate, m)
 	for _, t := range g.types {
+		tags, err := util.ParseClientGenTags(t.SecondClosestCommentLines)
```
I don't really like that we concentrate on SecondClosestCommentLines and ignore the comments below. At least we should show an error if our tags appear there.

deads2k commented:
/test pull-kubernetes-unit

deads2k commented:
/retest

caesarxuchao commented on staging/src/k8s.io/kube-gen/cmd/client-gen/generators/tags.go:
```diff
@@ -17,21 +17,9 @@ limitations under the License.
-
 // extractTag gets the comment-tags for the key.  If the tag did not exist, it
 // returns the empty string.
 func extractTag(key string, lines []string) string {
```
Move it to the util/tags.go, or remove it if it's no longer useful.

caesarxuchao commented:
@mfojtik could you send a follow up PR to address @sttts and my comments? I'm not removing the `lgtm` since I understand how frustrating it is to get through the flaky presubmits.

spxtr commented:
/lint

spxtr commented:
/lint

spxtr commented:
/lint

mfojtik committed:
regenerate clients

mfojtik commented:
@spxtr @sttts @deads2k @caesarxuchao i tried to address all Stefan comments during this rebase and also added missing godocs to fake clients... i guess we should land this asap and not grow the diff otherwise it will be nightmare to merge this ;-)

mfojtik commented:
/retest as github seems to be broken

deads2k commented:
/retest

deads2k commented:
@mfojtik doesn't build.

caesarxuchao commented on staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1beta1/apiservice.go:
```diff
@@ -57,6 +57,38 @@ func newAPIServices(c *ApiregistrationV1beta1Client) *aPIServices {
 	}
 }
 
+// Get takes name of the aPIService, and returns the corresponding aPIService object, and an error if there is any.
```
pkg/client/clientset_generated/clientset is removed in HEAD. Could you rebase?

mfojtik replied:
done, but still failing to compile...

caesarxuchao replied:
Sorry, i didn't notice this file is in k8s.io/aggregator. Github hid the path. Please keep it.

Please only remove `k8s.io/kubernetes/pkg/client/clientset_generated/clientset`, if there is any.

mfojtik commented:
@deads2k @sttts 

```
# k8s.io/kubernetes/vendor/k8s.io/client-go/informers/apps/v1beta2
vendor/k8s.io/client-go/informers/apps/v1beta2/scale.go:48: client.AppsV1beta2().Scales("k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".NamespaceAll).List undefined (type "k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes/typed/apps/v1beta2".ScaleInterface is interface with no methods)
vendor/k8s.io/client-go/informers/apps/v1beta2/scale.go:48: not enough arguments to return
	have (<T>)
	want (runtime.Object, error)
vendor/k8s.io/client-go/informers/apps/v1beta2/scale.go:51: client.AppsV1beta2().Scales("k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1".NamespaceAll).Watch undefined (type "k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes/typed/apps/v1beta2".ScaleInterface is interface with no methods)
vendor/k8s.io/client-go/informers/apps/v1beta2/scale.go:51: not enough arguments to return
```

i fixed the noMethods in types and regenerated stuff, but i'm getting this now...

caesarxuchao commented:
@mfojtik could you take a look why scale.go is generated? HEAD doesn't have any scale.go in informers/

mfojtik commented:
@caesarxuchao ^^ do you remember if scale.go was removed manually or something? i can't see the reason why that file is generated (the client is set to have no methods (iow. noVerbs)

mfojtik commented:
lol :-)

mfojtik commented:
@caesarxuchao ok I removed that file manually... seems like when I fixed the comment the file was not deleted automatically... lets see if the build pass now

mfojtik commented:
@deads2k @sttts green again \o/

sttts commented:
/lgtm

caesarxuchao commented:
I opened issue #49737 to ask for help for updating client-gen doc.

cc @droot @pmorie @MHBauer FYI, I remember you guys used client-gen in service-catalog. You'll need to update the tags in your types.go.
